File: E2EAppIOS/Sources/Screens/DialogScreen.swift
import SwiftUI

struct DialogScreen: View {
    @State private var result = "none"
    @State private var dialogOpen = false
    @State private var maybeCount = 0
    @State private var auto = Prefs.getBool("auto_dialog", false)

    var body: some View {
        ScreenColumn {
            TaggedText(tag: Tags.txtDialogResult, text: "dialog=\(result)")
            TaggedButton(tag: Tags.btnShowDialog, label: "ダイアログを開く") { dialogOpen = true }
            TaggedButton(tag: Tags.btnMaybeDialog, label: "交互にダイアログ") {
                maybeCount += 1
                // 乱数不使用: 奇数回目だけ開く決定的な交互動作が検証要件。
                if maybeCount % 2 == 1 { dialogOpen = true }
            }
            HStack {
                Text("起動時ダイアログ")
                Toggle("", isOn: $auto)
                    .labelsHidden()
                    .accessibilityIdentifier(Tags.swAutoDialog)
                    .onChange(of: auto) { newValue in Prefs.setBool("auto_dialog", newValue) }
            }
            TaggedText(tag: Tags.txtAutoDialog, text: "auto=\(auto ? "on" : "off")")
        }
        // auto=on のとき、この画面に入るたびダイアログを自動で開く。
        .onAppear { if auto { dialogOpen = true } }
        // UIAlertController は別ウィンドウに描画されるが、**ボタンには** accessibilityIdentifier が
        // そのまま届く(Android の testTagsAsResourceId のような再適用は不要)。
        // ただし title/message は UIAlertController が自前で描く StaticText で、
        // .accessibilityIdentifier を付けても捨てられる(実測。message 側に付けても同じ)。
        // → 見出しの検証はラベル「確認」で行う契約にする(#txt_dialog_title は iOS ネイティブには無い)。
        .alert("確認", isPresented: $dialogOpen) {
            Button("OK") {
                result = "ok"
            }
            .accessibilityIdentifier(Tags.btnDialogOK)
            Button("キャンセル", role: .cancel) {
                result = "cancel"
            }
            .accessibilityIdentifier(Tags.btnDialogCancel)
        }
    }
}

struct HealScreen: View {
    @State private var schemaV1 = Prefs.getBool("heal_schema_v1", true)
    @State private var tapped = "-"

    var body: some View {
        ScreenColumn {
            HStack {
                Text("旧ID(v1)を使う")
                Toggle("", isOn: $schemaV1)
                    .labelsHidden()
                    .accessibilityIdentifier(Tags.swHealSchema)
                    .onChange(of: schemaV1) { newValue in Prefs.setBool("heal_schema_v1", newValue) }
            }
            TaggedText(tag: Tags.txtHealSchema, text: "schema=\(schemaV1 ? "v1" : "v2")")

            // ラベル固定・id のみ切替がヒール検証の核: schema=v2 で id が解決不能でも、
            // ラベル「修復対象」から FM が着地できるかを見る。
            TaggedButton(tag: schemaV1 ? Tags.btnHealV1 : Tags.btnHealV2, label: "修復対象") {
                tapped = schemaV1 ? "v1" : "v2"
            }
            TaggedText(tag: Tags.txtHealResult, text: "tapped=\(tapped)")
            TaggedButton(tag: Tags.btnHealReset, label: "修復結果クリア") { tapped = "-" }
        }
    }
}

File: E2EAppCMP/composeApp/src/commonMain/kotlin/com/ftester/e2e/screens/DialogScreen.kt
package com.ftester.e2e.screens

import androidx.compose.foundation.layout.Row
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import com.ftester.e2e.Tags
import com.ftester.e2e.ui.ScreenColumn
import com.ftester.e2e.ui.TaggedButton
import com.ftester.e2e.ui.TaggedText
import com.ftester.e2e.util.exposeTestTagsAsResourceId
import com.ftester.e2e.util.Prefs

@Composable
fun DialogScreen() {
    var result by remember { mutableStateOf("none") }
    var dialogOpen by remember { mutableStateOf(false) }
    var maybeCount by remember { mutableStateOf(0) }
    var auto by remember { mutableStateOf(Prefs.getBool("auto_dialog", false)) }

    // auto=true のとき、この画面に入るたびダイアログを自動で開く。
    LaunchedEffect(Unit) {
        if (auto) dialogOpen = true
    }

    ScreenColumn(scrollable = true) {
        TaggedText(Tags.TXT_DIALOG_RESULT, "dialog=$result")
        TaggedButton(Tags.BTN_SHOW_DIALOG, "ダイアログを開く") { dialogOpen = true }
        TaggedButton(Tags.BTN_MAYBE_DIALOG, "交互にダイアログ") {
            maybeCount++
            // 乱数不使用: 奇数回目だけ開く決定的な交互動作が検証要件。
            if (maybeCount % 2 == 1) dialogOpen = true
        }
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text("起動時ダイアログ")
            Switch(
                checked = auto,
                onCheckedChange = {
                    auto = it
                    Prefs.setBool("auto_dialog", it)
                },
                modifier = Modifier.testTag(Tags.SW_AUTO_DIALOG)
            )
        }
        TaggedText(Tags.TXT_AUTO_DIALOG, "auto=${if (auto) "on" else "off"}")
    }

    if (dialogOpen) {
        // ダイアログは**別ウィンドウ**に描画されるため App() ルートの
        // exposeTestTagsAsResourceId() が効かない。ここで再適用しないと Android では
        // ダイアログ内の testTag が resource-id 化されず #id を一切引けない(ラベルだけ残る)。
        AlertDialog(
            modifier = Modifier.exposeTestTagsAsResourceId(),
            onDismissRequest = { dialogOpen = false },
            title = { TaggedText(Tags.TXT_DIALOG_TITLE, "確認") },
            confirmButton = {
                TaggedButton(Tags.BTN_DIALOG_OK, "OK") {
                    result = "ok"
                    dialogOpen = false
                }
            },
            dismissButton = {
                TaggedButton(Tags.BTN_DIALOG_CANCEL, "キャンセル") {
                    result = "cancel"
                    dialogOpen = false
                }
            }
        )
    }
}
